### DISCUSSION TITLE
AttributeError: 'HeteroDataBatch' has no attribute 'y'

### DISCUSSION THREAD

**Discussion by AlizadehF** [pyg-team/pytorch_geometric#8828] [Q&A]
Hello, 
my code is> 
model.train()
    for batch in train_loader: 
        out = model(batch.x_dict, batch.edge_index_dict, batch)  
        
        loss = criterion(out, batch.y) 
        loss.backward()  
        optimizer.step()  
        optimizer.zero_grad()  
error is>
AttributeError: 'HeteroDataBatch' has no attribute 'y'

and when used batch.y_dict, the error is >
TypeError: cross_entropy_loss(): argument 'target' (position 2) must be Tensor, not dict

Can you make suggestions for this error?

**Comment by wsad1** [comment]
Could you share the result of `print(batch)`. 
I am guessing `y` is a node level tensor, in that case you have to pass `batch.y_dict['node_type']` to `criterion`, where `node_type` is the node type you are making a prediction on.

**Reply by AlizadehF** [reply]
print (batch):

`HeteroDataBatch(
  y=[763144],
  userId={
    x=[763144, 1],
    batch=[763144],
    ptr=[2]
  },
  processId={
    x=[763144, 1],
    batch=[763144],
    ptr=[2]
  },
  eventId={
    x=[763144, 1],
    batch=[763144],
    ptr=[2]
  },
  (userId, processName, processId)={
    edge_index=[2, 763144],
    edge_attr=[1]
  },
  (userId, eventName, eventId)={
    edge_index=[2, 763144],
    edge_attr=[1]
  },
  (processId, rev_processName, userId)={ edge_index=[2, 763144] },
  (eventId, rev_eventName, userId)={ edge_index=[2, 763144] },
  (userId, rev_processName, processId)={ edge_index=[2, 763144] },
  (userId, rev_eventName, eventId)={ edge_index=[2, 763144] },
  (processId, rev_processName, userId)={ edge_index=[2, 763144] },
  (eventId, rev_eventName, userId)={ edge_index=[2, 763144] },
  (userId, rev_processName, processId)={ edge_index=[2, 763144] },
  (userId, rev_eventName, eventId)={ edge_index=[2, 763144] },
  (processId, rev_processName, userId)={ edge_index=[2, 763144] },
  (eventId, rev_eventName, userId)={ edge_index=[2, 763144] }`

**Comment by AlizadehF** [comment]
Thanks
I have hetero edges with hetero label

And each category of edge have different y size.
In the out I have tensor with size 162, 162, 55,55 but in the batch.y I
have size 162, 55, 162, 55

**Comment by AlizadehF** [comment]
It means I have different size of label for each subgraph or edge type

I have an edge classification Task. i have different node type, different edge type and for each edge, 22 features and label i have.
for example, 

```
data = HeteroData()
data['n1'].x = torch.tensor(df['f1']).reshape([-1, 1])
data['n1'].x = data['n1'].x.type(torch.FloatTensor)
data['n2'].x = torch.tensor(df['f2']).reshape([-1, 1])
data['n2'].x = data['n2'].x.type(torch.FloatTensor)
data['n1','type1', 'n2'].edge_index = e1
data['n1','type1', 'n2'].edge_attr = node_features

data['n3'].x = torch.tensor(df['f3']).reshape([-1, 1])
data['n3'].x = data['n3'].x.float()
data['n4'].x = torch.tensor(df['f3']).reshape([-1, 1])
data['n4'].x = data['n4'].x.float()
data['n3','type2', 'n4'].edge_index = e3
data['n3','type2', 'n4'].edge_attr = node_features

data['n1','type1', 'n2'].y = y1
data['n3','type2', 'n4'].y = y3
```
my label vector size is different per edge type

**Comment by AlizadehF** [comment]
`
data = T.ToUndirected()(data)
class GNN(torch.nn.Module):
    def __init__(self, hidden_channels, out_channels):
        super().__init__()
        self.conv1 = SAGEConv((-1,-1), hidden_channels)
        self.conv2 = SAGEConv((-1,-1), out_channels)

    def forward(self, x, edge_index):
        x = self.conv1(x, edge_index).relu()
        x = self.conv2(x, edge_index)
        return x

model = GNN(hidden_channels=64, out_channels=1)
model = to_hetero(model, data.metadata(), aggr='sum')
`

`
out = model(data.x_dict, data.edge_index_dict)
loss = criterion(out, data.y_dict)  
loss.backward(retain_graph=True)  
optimizer.step()  
optimizer.zero_grad()  
`

error:
TypeError: cross_entropy_loss(): argument 'input' (position 1) must be Tensor, not dict

i guess my model is for node classification but my goal is edge classification, for this aim, the `out` is for nodes and the `data.y_dict` is for edges.
i dont know, how to make a GNN mode for edge classification Heterogeneous Graph.

**Reply by rusty1s** [reply]
Please note that GNNs compute node-level representations. The general procedure to let them return edge-level representations (e.g., used for edge classification), is to combine the source and destination node features and pass them to a classifier, e.g.:
```python
class MyModel():
    def __init__(self., ...):
        self.gnn = to_hetero(MyGNN())
        self.classifier = Linear(2 * hidden_channels, num_classes)
    
    def forward(self, x_dict, edge_index_dict):
        x_dict = self.gnn(x_dict, edge_index_dict)
        x_src = x_dict[src_node_type]
        x_dst = x_dict[dst_node_type]
        return self.classifier(torch.cat([x_src, x_dst], dim=-1))

**Reply by AlizadehF** [reply]
thanks for your response,, 
i define this code x_src = x_dict[src_node_type], x_dst = x_dict[dst_node_type] for all node types?
and , this code is for edge classification??

**Reply by rusty1s** [reply]
You can define them for all edge types you are interested in for classification

**Reply by AlizadehF** [reply]
I have used this code for __init__
`def __init__(self, hidden_channels, out_channels):
        super().__init__()
        self.conv1 = SAGEConv((-1,-1), hidden_channels)
        self.conv2 = SAGEConv((-1,-1), hidden_channels)
        self.pool = aggr.MeanAggregation()
        self.lin = Linear(hidden_channels, 11)`

your __init__ def is different , may you explain `self.classifier = Linear(2 * hidden_channels, num_classes)` ?

**Reply by rusty1s** [reply]
So in my example, you would have a GNN (just as you did), but then have a parent module that wraps the heterogeneous GNN (via `to_hetero()`) and a final edge-level predictor.3
